bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java

```diff
@@ -0,0 +1,714 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.gree.internal.handler;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.gree.internal.GreeCryptoUtil;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.gson.GreeBindRequest4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeBindRequestPack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeBindResponse4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeBindResponsePack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeExecCommand4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeExecResponse4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeExecResponsePack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeExecuteCommandPack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeReqStatus4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeReqStatusPack4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeScanResponse4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeStatusResponse4GsonDTO;
+import org.openhab.binding.gree.internal.gson.GreeStatusResponsePack4GsonDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The GreeDevice object repesents a Gree Airconditioner and provides
+ * device specific attributes as well a the functionality for the Air Conditioner
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+// @NonNullByDefault

```

Comment:
I see declaring this as NonNullByDefault has bigger impact. Nevertheless, I'd like to encourage you to use this opportunity to make your code even more robust.

Reply:
ok, I worked on that, please check the result